iceoryx_hoofs/include/iceoryx_hoofs/design_pattern/polymorphic_handler.hpp

```diff
@@ -0,0 +1,126 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_DESIGN_PATTERN_POLYMORPHIC_HANDLER_HPP
+#define IOX_HOOFS_DESIGN_PATTERN_POLYMORPHIC_HANDLER_HPP
+
+#include <atomic>
+#include <iostream>
+#include <type_traits>
+
+#include "iceoryx_hoofs/design_pattern/static_lifetime_guard.hpp"
+
+namespace iox
+{
+namespace design_pattern
+{
+namespace detail
+{
+
+/// @brief default hooks for the PolymorphicHandler
+/// @tparam Interface the handler interface
+/// @note template hooks to avoid forced virtual inheritance
+template <typename Interface>
+struct DefaultHooks
+{
+    /// @brief called if the polymorphic handler is set or reset after finalize
+    /// @param currentInstance the current instance of the handler singleton
+    /// @param newInstance the instance of the handler singleton to be set
+    static void onSetAfterFinalize(Interface& currentInstance, Interface& newInstance) noexcept;
+};
+
+} // namespace detail
+
+/// @brief Implements a singleton handler that has a default instance and can be changed
+/// to another instance at runtime. All instances have to derive from the same interface.
+/// The singleton handler owns the default instance but all other instances are created externally.
+/// @tparam Interface The interface of the handler instances.
+/// @tparam Default The type of the default instance. Must be equal to or derive from Interface.
+/// @tparam Hooks A struct that implements onSetAfterFinalize which is called when
+/// attempting to set or reset the handler after finalize was called.
+///
+/// @note In the special case where Default equals Interface, no polymorphism is required.
+/// It is then possible to e.g. switch between multiple instances of Default type.
+/// @note The lifetime of external non-default instances must exceed the lifetime of the PolymorphicHandler.
+/// @note The PolymorphicHandler is guaranteed to provide a valid handler during the whole program lifetime (static).
+/// It is hence not advisable to have other static variables depend on the PolymorphicHandler.
+/// It must be ensured that they are destroyed before the PolymorphicHandler.
+/// @note Hooks must implement
+/// static void onSetAfterFinalize(Interface& /*currentInstance*/, Interface& /*newInstance*/).
+template <typename Interface, typename Default, typename Hooks = detail::DefaultHooks<Interface>>
+class PolymorphicHandler
+{
+    static_assert(std::is_base_of<Interface, Default>::value, "Default must inherit from Interface");
```

Comment:
A class is always base of itself (as intended), so `PolymorphicHandler<Default, Default>` works. The consequence is that it is not really polymorphic anymore, but you still can switch between different `Default` instances. That is a niche feature but not detrimental and potentially useful.

Reply:
You misunderstood me. My point was that `Default` can be seen as the base of itself but I wouldn't say the `Default` inherited from itself. The message implies that this is the case, therefore my proposal to change the wording and use `base` instead of `inherit`.